### PR TITLE
Kafka

### PR DESCRIPTION
**Pull request by ajax-butriy-v**

### CODE REVIEW

File: internal-api/src/main/proto/com/example/internal/output/pubsub/pass/transferred_pass_message.proto [hunk 1]
```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
+
+package com.example.internal.input.reqreply;
+
+import "com/example/internal/commonmodels/pass.proto";
+
```
Comment (ajax-torbin-v) [inline internal-api/src/main/proto/com/example/internal/output/pubsub/pass/transferred_pass_message.proto:6]: Needless blank line(s) [NoConsecutiveBlankLines]

File: pass-manager-svc/build.gradle.kts [hunk 2]
```diff
@@ -12,8 +12,12 @@ dependencies {
     implementation(libs.mongock.springboot.v3)
     implementation(libs.mongock.mongodb.springdata.v4.driver)
     implementation(libs.nats)
+    implementation("org.springframework.kafka:spring-kafka:3.2.4")
+    implementation("io.projectreactor.kafka:reactor-kafka:1.3.23")
+
```
Comment (ajax-torbin-v) [inline pass-manager-svc/build.gradle.kts]: Needless blank line(s) [NoConsecutiveBlankLines]

File: gateway/src/main/kotlin/com/example/gateway/web/mapper/proto/pass/CreatePassResponseMapper.kt [hunk 3]
```diff
@@ -21,11 +21,9 @@ object CreatePassResponseMapper {
     }
 
     fun PassDto.toCreatePassRequest(): CreatePassRequest {
-        return CreatePassRequest.newBuilder()
-            .setPassTypeId(passTypeId)
-            .setPassOwnerId(passOwnerId)
-            .setPurchasedFor(purchasedFor.toBDecimal())
-            .build()
+        return CreatePassRequest.newBuilder().also {
+            it.passToCreate = this.toProto()
```
Comment (ajax-malov-n) [inline gateway/src/main/kotlin/com/example/gateway/web/mapper/proto/pass/CreatePassResponseMapper.kt]: remove this

File: pass-manager-svc/build.gradle.kts [hunk 4]
```diff
@@ -12,8 +12,12 @@ dependencies {
...
     implementation(project(":core"))
     implementation(project(":internal-api"))
     testImplementation(libs.nats.embedded)
+    testImplementation("org.springframework.kafka:spring-kafka-test:3.2.4")
```
Comment (ajax-malov-n) [inline pass-manager-svc/build.gradle.kts]: Do you need this?

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImpl.kt [hunk 5]
```diff
@@ -23,4 +33,35 @@ internal class PassOwnerStatisticsServiceImpl(
     override fun calculatePriceDistributions(passOwnerId: String): Flux<PriceDistribution> {
         return passRepository.getPassesPriceDistribution(passOwnerId)
     }
+
+    override fun publishTransferPassStatistics(pass: MongoPass, previousPassOwnerId: String): Mono<Unit> {
+        return passTypeService.getById(pass.passTypeId.toString())
+            .map { passType -> mapToStatisticsWithPassTypeId(passType, pass, previousPassOwnerId) }
+            .flatMap { (message, passTypeId) ->
+                transferPassStatisticsMessageProducer.sendTransferPassStatisticsMessage(message, passTypeId)
+            }
```
Comment (ajax-torbin-v) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImpl.kt]: If message contatins passTypeId, you can use it without creating pair

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/producer/TransferPassMessageProducer.kt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    fun sendTransferPassMessage(updatedPass: MongoPass, key: String, previousOwnerId: String): Mono<Unit> {
+        val senderRecord = SenderRecord.create(
+            ProducerRecord(
+                KafkaTopic.KafkaTransferPassEvents.TRANSFER,
+                key,
+                updatedPass.toTransferredPassMessage(previousOwnerId).toByteArray()
+            ),
+            null
+        )
+        return kafkaSender.send(senderRecord.toMono()).then(Unit.toMono())
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/producer/TransferPassMessageProducer.kt:25]: Prefer to write code from top to bottom

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/nats/controller/pass/CreatePassNatsController.kt [hunk 7]
```diff
@@ -28,13 +28,15 @@ class CreatePassNatsController(
...
-            .onErrorResume {
-                logger.error("Error occurred while executing", it)
-                failureCreatedPassResponse(it).toMono()
-            }
+        return request.passToCreate.let { protoPass ->
+            passService.create(protoPass.toModel(), protoPass.passOwnerId, protoPass.passTypeId)
+                .map { createdPass -> createdPass.toProto() }
+                .map { proto -> proto.toSuccessCreatePassResponse() }
+                .onErrorResume {
+                    logger.error("Error occurred while executing", it)
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/nats/controller/pass/CreatePassNatsController.kt]: Make more meaningful exception message

File: pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImplTest.kt [hunk 8]
```diff
@@ -67,4 +82,53 @@ internal class PassOwnerStatisticsServiceImplTest {
...
+
+        // WHEN
+        val actual = passOwnerStatisticsService.mapToStatisticsWithPassTypeId(
+            passTypeWithNullPrice,
+            passWithNullPurchasedFor,
+            previousPassOwnerId
+        )
+
+        // THEN
+        assertThat(actual.first.wasPurchasedWithDiscount).isFalse()
```
Comment (ajax-torbin-v) [inline pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImplTest.kt]: assertFalse?

File: pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/util/IntegrationTest.kt [hunk 9]
```diff
@@ -1,8 +1,24 @@
...
+@MockkBean(
+    relaxed = true,
+    value = [
+        TransferPassMessageConsumer::class,
+        TransferPassStatisticsMessageProducer::class,
+        TransferPassMessageProducer::class,
+        KafkaReactiveConsumerConfiguration::class,
+        KafkaReactiveProducerConfig::class
+    ]
+)
```
Comment (ajax-torbin-v) [inline pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/util/IntegrationTest.kt]: annotation?

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImpl.kt [hunk 10]
```diff
@@ -23,4 +33,35 @@ internal class PassOwnerStatisticsServiceImpl(
...
+        val passTypePrice = passType.price ?: BigDecimal.ZERO
+        val deltaBetweenPassAndPassTypePrices = passTypePrice - (pass.purchasedFor ?: BigDecimal.ZERO)
+        val isDeltaPositive = deltaBetweenPassAndPassTypePrices.signum() > 0
+
+        val transferStatisticsMessage = pass.toTransferPassStatisticsMessage(
+            passType = passType,
+            previousOwnerId = previousOwnerId,
+            isDeltaPositive = isDeltaPositive
+        )
+        log.info("Statistics message {}", transferStatisticsMessage)
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImpl.kt]: Do you need this?

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImpl.kt [hunk 11]
```diff
@@ -23,4 +33,35 @@ internal class PassOwnerStatisticsServiceImpl(
...
+    override fun publishTransferPassStatistics(pass: MongoPass, previousPassOwnerId: String): Mono<Unit> {
+        return passTypeService.getById(pass.passTypeId.toString())
+            .map { passType -> mapToStatisticsWithPassTypeId(passType, pass, previousPassOwnerId) }
+            .flatMap { (message, passTypeId) ->
+                transferPassStatisticsMessageProducer.sendTransferPassStatisticsMessage(message, passTypeId)
+            }
+            .thenReturn(Unit)
+    }
+
+    fun mapToStatisticsWithPassTypeId(
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/service/impl/PassOwnerStatisticsServiceImpl.kt]: Maybe rename method to `create...TypeId`

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/web/mapper/proto/pass/TransferredPassStatisticsMessageMapper.kt [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+import java.time.Instant
+import com.google.protobuf.Duration as ProtoDuration
+
+object TransferredPassStatisticsMessageMapper {
+    fun MongoPass.toTransferPassStatisticsMessage(
+        passType: MongoPassType,
+        previousOwnerId: String,
+        isDeltaPositive: Boolean,
+    ): TransferredPassStatisticsMessage {
+        val duration = Duration.between(Instant.now(), passType.activeTo)
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/web/mapper/proto/pass/TransferredPassStatisticsMessageMapper.kt]: Logic in mapper

File: pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/util/IntegrationTest.kt [hunk 13]
```diff
@@ -1,8 +1,24 @@
...
 abstract class IntegrationTest
```
Comment (ajax-malov-n) [inline pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/util/IntegrationTest.kt]: Provide more meaningful name or Mock where it needed

File: pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/consumer/TransferPassMessageConsumer.kt [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    private fun delegateToStatisticsService(record: ReceiverRecord<String, ByteArray>): Mono<Unit> {
+        return recordToPassAndPreviousOwnerIdPair(record)
+            .flatMap { (updatedPass, previousOwnerId) ->
+                passOwnerStatisticsService.publishTransferPassStatistics(
+                    updatedPass,
+                    previousOwnerId
+                )
+            }
+            .doFinally { record.receiverOffset().acknowledge() }
+    }
```
Comment (ajax-shulepa-a) [inline pass-manager-svc/src/main/kotlin/com/example/passmanagersvc/kafka/consumer/TransferPassMessageConsumer.kt]: this method looks like business logic to me. refactor it to not operate with infrastructure details like ReceiverRecord

File: pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/kafka/consumer/TransferPassMessageConsumerTest.kt [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+@ActiveProfiles("test")
+class TransferPassMessageConsumerTest {
+    @Autowired
+    private lateinit var transferPassMessageProducer: TransferPassMessageProducer
+
+    @Autowired
+    private lateinit var reactiveMongoTemplate: ReactiveMongoTemplate
+
+    @Autowired
+    private lateinit var consumerForTransferStatistics: KafkaReceiver<String, ByteArray>
```
Comment (ajax-shulepa-a) [inline pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/kafka/consumer/TransferPassMessageConsumerTest.kt:39]: is this bean from main code? is it ok to use it to subscribe in tests?

Comment (ajax-butriy-v) [reply pass-manager-svc/src/test/kotlin/com/example/passmanagersvc/kafka/consumer/TransferPassMessageConsumerTest.kt:39]: No, it is consumer from ConsumerForSecondTopic class, which is nested class in this test :)
I use it to verify that my Kafka flow is actually working by checking there are valid messages in second topic

**Review by ajax-shulepa-a — APPROVED** [review]

**Pull request merged**